feat(Modal): added support for React Portal

fantua opened the pull request:
- Added `react-portal` package (uses React v16 and its official API for creating portals; has a fallback for React v15).
- Updated `enzyme-adapter-react-16` to latest version for `Portals` support.

1. Currently `handleBackdropClick` is fires for nested modals and close all opened. Need to investigate more on this. 
2. Most test is failing by: 
```js
expect(wrapper.childAt(0).children().length).toBe(0);
```
After using `ReactDOM.createPortal` it's not more 0. Can I remove this checks?

Need some reviews/comments to current changes.

Refers to #601, #761.

fantua commented:
cc @TheSharpieOne @supergibbs

supergibbs commented:
I haven't had a chance to mess around with portals yet but would it be easier to just wrap a `Modal` in another component in your app and tell it to render to a portal? Since reactstrap is still supports react 15 not sure if it should have react 16 stuff built in.

fantua commented:
> I haven't had a chance to mess around with portals yet but would it be easier to just wrap a Modal in another component in your app and tell it to render to a portal?

It's doesn't make sense because currently `Modal` uses `unstable_renderSubtreeIntoContainer`. `renderSubtreeIntoContainer` isn't stable and have some bugs (ex. #601), my changes allow you to:
- use `Portal` for 16 (which is stable)
- fix component logic and structure (currently it's looks a little bit weird)
- leaves fallback for 15

seberik commented:
Calling destroy in componenDidUpdate breaks the modal close transition.

TheSharpieOne commented:
Exit animations are not working
Closing a nested model closes all of the modals. (not even clicking the backdrop)
For the tests, you can change them however you need to; just ensure they have good/complete coverage and are actually testing (not just running through the code to get coverage) and pass.

fantua commented:
Added new changes:
- fixed Modal exit animation,
- fixed nested Modals,
- updated react-transition-group to 2.2.1

@seberik @TheSharpieOne, guys take a look at this, looks like now all should be good 😃

seberik commented:
The exit animation now works but when double-clicking the backdrop or escape-key it will close and re-open the modal.

fantua commented:
Fixed backdrop  double-clicking and tests.

fantua commented:
Merged with master. Waiting for review.

cc @TheSharpieOne

TheSharpieOne commented on src/__tests__/Modal.spec.js:
```diff
@@ -76,7 +76,7 @@ describe('Modal', () => {
 
     jest.runTimersToTime(300);
-    expect(wrapper.childAt(0).children().length).toBe(0);
+    expect(wrapper.childAt(0).children().length).not.toBe(0);
```
You can remove all of the checks like this. The previous test was ensuring nothing was inserted into the DOM, but now using react-portal, we will always be inserting react-portal, so these checks are more/less pointless (unless you want to check that react-portal get inserted, but I would also only do that in one test, not all of them.)

fantua replied:
Removed those unneeded checks and created new test for this.
